### PR TITLE
Update VariantLocation hierarchy to Association hierarchy, using GenomicLocationAssociation

### PR DESCRIPTION
**Pull request by sierra-moxon**
Requirements: merge or otherwise reuse core:GenomicLocationAssociation in Variant schema.

- VariantLocation becomes a mixin
- Variant[Genomic|Transcript|Protein]Location -> Variant[Genomic|Transcript|Protein]LocationAssociation 
- added VariantDTO and test data (updated basic Variant test data and added test data for the LocationAssociations above).
- because I added variant-specific entities, I needed to loosen the requirements for the generic GenomicLocationAssociation class (e.g. genomic_entity_curie is no longer required in the base class, but `variant_entity_curie` is_a: `genomic_entity_curie` and is required).  
  - I also had to remove 'required' from has_assembly, but no classes were using that GenomicLocationAssociation yet as far as I could tell.  I also decided to make this abstract to help promote using it as a base class and further restricting the content in individual subclasses.

### CODE REVIEW

File: model/schema/variantDTO.yaml [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+classes:
+
+  VariantDTO:
+    is_a: GenomicEntityDTO
+    description: >-
+      Ingest class for an Allele object
+    slots:
+      - variant_type
+      - related_notes
+      - source_general_consequence
```
Comment (gildossantos) [inline model/schema/variantDTO.yaml:47]: `consequence` and `source_general_consequence` have domain `VariantLocation` (the mixin), but are here listed as slots for the `VariantDTO` (which does not have and probably should not have this mixin). I would've expected this to break things? Maybe the DTO objects are outside of some checks (like the `make` command that builds the big `json` and `yaml` files)?

Comment (gildossantos) [reply model/schema/variantDTO.yaml:47]: Looking at non-DTO variant file, consequence is only in the location associations. Should we remove consequence from the VariantDTO then, but leave it in various location association DTOs?

File: model/schema/variantDTO.yaml [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+      - assembly_curie
+      - start
+      - end
+      - exon_number
+      - intron_number
+      - variation_strand
+      - number_additional_dna_base_pairs
+      - number_removed_dna_base_pairs
+      - inserted_sequence
+      - deleted_sequence
```
[the remark this answers is not in the document]
Comment (gildossantos) [reply model/schema/variantDTO.yaml:71]: Similar question for `reference_sequence` slot value, which comes from the mixin.

**Review by gildossantos — COMMENTED** [review]
Looks good to me overall. I don't think I found any breaking issues, but lots of places where I could benefit from some more documentation perhaps?

**Comment by sierra-moxon** [comment]
Thanks for the review @gildossantos!  Lots of good stuff in the review; will add to this PR.

File: model/schema/core.yaml [hunk 3]
```diff
@@ -371,10 +367,15 @@ classes:
       - subject
```
[the remark this answers is not in the document]
Comment (sierra-moxon) [reply model/schema/core.yaml:367]: These are in the defining slots section, I think they are ok here - they basically give a clue to the user about what constitutes a valid object.

File: model/schema/variantDTO.yaml [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+  VariantDTO:
+    is_a: GenomicEntityDTO
+    description: >-
+      Ingest class for an Allele object
+    slots:
+      - variant_type
+      - related_notes
+      - source_general_consequence
+      - variant_status
+      - reference_sequence
```
[the remark this answers is not in the document]
Comment (sierra-moxon) [reply model/schema/variantDTO.yaml:49]: reference_sequence is an actual sequence of the non-variant span of the sequence, the sequence_of_reference_accession_number is an accession number for the assembly and chromosome of the reference sequence. 

I think the definitions are pretty good here; is there anything that should change?

File: test/data/variant_association_ingest_test.json [hunk 5]
```diff
@@ -0,0 +1,72 @@
+{
+  "linkml_version": "1.7.3",
+  "variant_genomic_location_association_ingest_set": [
+    {
+      "variant_entity_curie": "ZFIN:ZDB-ALT-000412-8",
+      "predicate_name": "has_genomic_location",
+      "chromosome_curie": "7",
```
[the remark this answers is not in the document]
Comment (sierra-moxon) [reply test/data/variant_association_ingest_test.json:7]: I think this is a good question for discussion in the variant WG; `chromosome_curie` and `assembly_curie` are both following naming conventions for the schema, but I don't think we actually have requirements to make them curies.  

I added "display_name" to the Chromosome class.

File: model/schema/variation.yaml [hunk 6]
```diff
@@ -124,9 +129,10 @@ classes:
       - exon_number
```
[the remark this answers is not in the document]
Comment (sierra-moxon) [reply model/schema/variation.yaml:129]: Good catch on the object -- fixed by removing transcript and polypeptide as individual slots and moving their ranges to slot_usage.

`has_genomic_location` predicate is another question for the variant WG.

File: model/schema/variantDTO.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (sierra-moxon) [reply model/schema/variantDTO.yaml:70]: good question for the group; yep.

File: model/schema/variantDTO.yaml [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+      - end
+      - exon_number
+      - intron_number
+      - variation_strand
+      - number_additional_dna_base_pairs
+      - number_removed_dna_base_pairs
+      - inserted_sequence
+      - deleted_sequence
+      - padded_base
+      - dna_mutation_type
```
[the remark this answers is not in the document]
Comment (sierra-moxon) [reply model/schema/variantDTO.yaml:72]: dna_mutation_type is a more specific SOTerm than those used in variant_type.  e.g. 'variant_type' would be something like 'point_mutation' (SO:1000008), the 'dna_mutation_type' would be 'T_to_C_transition' (SO:1000013). 
I edited the description a bit, and noted that this is for ZFIN.

**Pull request merged**